File: re/lol/lol.py
"""
Introducing the linked object list (.lol) format.

A .lol file consists of three sections:
- The length of the reconstructed file, as an 8-byte BE unsigned integer.
- The MD5 hash of the reconstructed file, as a checksum.
- A sequence of variable-length chunks, described below.

A "chunk" itself consists of three sections:
- The length of the data part of the chunk, as a 4-byte BE unsigned integer
(that is, it excludes the length and the pointer at the end of the chunk).
- The data contained in the chunk.
- A pointer to the next chunk to read, expressed as an 8-byte BE unsigned integer
of the absolute starting offset of the next chunk. If set to all null bytes,
this indicates the "end" of the file (at which point the reconstructed file's hash
and length should be checked).

The first chunk must always be the first sequential chunk (which will hopefully
make things more obvious?). The last chunk may appear anywhere in the file.

Of course, you could just do this where each chunk is sequential and equal-size,
but where's the fun in that?
"""

import argparse
import hashlib
import logging
import random
from pathlib import Path
from typing import Optional, List

DEFAULT_INPUT_FILE = "flag.png"

# Anywhere from 1 to 10 kiB
DEFAULT_MIN_RANDOM_CHUNK_SIZE = 1024
DEFAULT_MAX_RANDOM_CHUNK_SIZE = DEFAULT_MIN_RANDOM_CHUNK_SIZE * 10


FORMAT = "[%(levelname)s] %(filename)s:%(lineno)s - %(funcName)s(): %(message)s"
logging.basicConfig(format=FORMAT)
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)


class Chunk:
    """
    Format:
    - length of data (4 bytes, BE unsigned int)
    - data
    - absolute offset to next chunk (8 bytes, BE unsigned int)
    """

    # Constants, in bytes
    SIZE_CHUNK_LENGTH = 4
    SIZE_CHUNK_OFFSET = 8

    def __init__(self, data: bytes, idx: int):
        """
        Create a "raw" uninitialized chunk.

        :param data: The raw binary data of the chunk.
        :param idx: The index of this chunk in a larger (correct) sequence of
            chunks.
        """
        self.data: bytes = data
        self.idx: int = idx
        self.next: Optional[int] = None

        self.is_final_chunk: bool = False

        if not self._check_len():
            raise RuntimeError(
                "Chunk length exceeds UINT32_MAX, the maximum size of a chunk"
            )

    @classmethod
    def get_next_chunk_data(cls, data: bytes, cur_offset: int) -> "bytes, int":
        """
        Return a tuple containing the data of the chunk starting at `cur_offset`
        and the offset of the next chunk.
        """
        start_offset = cur_offset
        data_offset = start_offset + cls.SIZE_CHUNK_LENGTH

        chunk_length = int.from_bytes(data[cur_offset:data_offset], "big")
        next_offset = data_offset + chunk_length

        chunk_data = data[data_offset:next_offset]
        next_pointer = int.from_bytes(
            data[next_offset : next_offset + cls.SIZE_CHUNK_OFFSET], "big"
        )

        return chunk_data, next_pointer

    def _check_len(self) -> bool:
        """
        Assert that the length of the data is not greater than UINT32_MAX.

        Return True if it is within the allowed range, False otherwise.
        """
        return len(self.data) <= (2 ** (8 * self.SIZE_CHUNK_LENGTH)) - 1

    def as_bytes(self) -> bytes:
        """
        Get the full contents of this chunk as bytes.

        If the object is uninitialized, as indicated by `next` being
        None, this raises RuntimeError.
        """
        if self.next is None:
            raise RuntimeError("Attempted to get the bytes of an uninitialized chunk")

        return (
            len(self.data).to_bytes(self.SIZE_CHUNK_LENGTH, "big")
            + self.data
            + self.next.to_bytes(self.SIZE_CHUNK_OFFSET, "big")
        )

    def get_full_length(self) -> bytes:
        """
        Get the full length of this chunk.
        """
        return len(self.data) + self.SIZE_CHUNK_LENGTH + self.SIZE_CHUNK_OFFSET


class LOLFile:
    """
    Format:
    - length of reconstructed file (8 bytes, BE unsigned int)
    - md5 hash of reconstructed file (16 bytes)
    - list of chunks (first chunk is always start of reconstructed file)
    """

    # Constants, in bytes
    SIZE_FILE_LENGTH = 8
    SIZE_MD5_LENGTH = 16

    @staticmethod
    def get_raw_chunks_from_file(
        file_path: Path, min_chunk_size: int, max_chunk_size: int, seed=None
    ) -> List[Chunk]:
        """
        Create a list of randomly-sized chunks from a file.

        The raw chunks are returned in sequential order, such that concatenating
        the `data` attribute of the resulting list of chunks would produce the
        original file.

        If `min_chunk_size` is greater than `max_chunk_size`, this raises
        ValueError.

        :param file_path: The path to the file to break into chunks.
        :param min_chunk_size: The minimum size of the `data` field of a chunk.
        :param max_chunk_size: The maximum size of the `data` field of a chunk.
        :param seed: The seed to use when generating chunk sizes.
        :returns: A list of Chunk objects corresponding to the contents of the input
            file.
        """
        if max_chunk_size < min_chunk_size:
            raise ValueError("Max raw chunk size is smaller than min raw chunk size")

        # The same seed will always lead to the same random outputs.
        randomizer = random.Random(seed)

        # List of chunks.
        raw_chunks: List[Chunk] = []

        # Read in a random amount of the file. So long as we get bytes back,
        # continue attempting to read.
        with open(args.input_file, "rb") as fp:
            idx = 0
            while data := fp.read(randomizer.randint(min_chunk_size, max_chunk_size)):
                raw_chunks.append(Chunk(data, idx))
                idx += 1

        # Indicate that the final chunk should have its `next` pointer set to null
        raw_chunks[-1].is_final_chunk = True

        logger.debug(f"Created {len(raw_chunks)} chunks")
        logger.debug(f"Chunk data lengths: {[len(chunk.data) for chunk in raw_chunks]}")

        return raw_chunks

    @classmethod
    def from_chunks(cls, raw_chunks: List[Chunk], seed=None) -> bytes:
        """
        Construct a LOLFile.

        :param raw_chunks: A sequence of chunks with data in the correct order.
            That is, if the data were concatenated as-is from the list of chunks,
            the original file would pop out.
        :param seed: A valid seed for `random.Random`. If None, the seed is random
            based on the implementation of the `random` library.
        """
        # The same seed will always lead to the same random outputs.
        randomizer = random.Random(seed)

        # Create a dictionary from the raw chunks, so that we know the original
        # order of the chunks.
        chunk_indexes = {idx: chunk for idx, chunk in enumerate(raw_chunks)}

        # Randomize the order of the chunks (note that shuffle() is in-place,
        # which we don't want since we need the original chunks). Always
        # keep the first "true" chunk first.
        random_chunks: list[Chunk] = [raw_chunks[0]]
        random_chunks += randomizer.sample(raw_chunks[1:], len(raw_chunks) - 1)

        # Calculate the offset of each ordered chunk into the resulting file.
        random_offsets = {}
        cur_offset = cls.SIZE_FILE_LENGTH + cls.SIZE_MD5_LENGTH
        for chunk in random_chunks:
            random_offsets[chunk.idx] = cur_offset
            cur_offset += chunk.get_full_length()

        # With the offsets of each correctly-ordered chunk known, now link up
        # each chunk's "next" pointer
        for chunk in random_chunks:
            if chunk.idx + 1 not in random_offsets:
                logger.debug(f"Assuming chunk ending in {chunk.data[-20:]} is the end")
                chunk.next = 0
            else:
                chunk.next = random_offsets[chunk.idx + 1]

        # Get the information we need for the header
        raw_bytes = b"".join([chunk.data for chunk in raw_chunks])
        original_len = len(raw_bytes)
        md5_hash = hashlib.md5(raw_bytes).digest()

        # Return full sequence of bytes
        return (
            original_len.to_bytes(cls.SIZE_FILE_LENGTH, "big")
            + md5_hash
            + b"".join([chunk.as_bytes() for chunk in random_chunks])
        )

    @classmethod
    def undo_lol_file(cls, lol_file: bytes) -> bytes:
        """
        Undo and verify a .lol file.

        If the file fails the length and hash check, this raises RuntimeError.
        """
        len_bytes = int.from_bytes(lol_file[0:8], "big")
        md5_bytes = lol_file[8:24]

        reconstructed_data = b""
        cur_offset = 24
        while True:
            chunk_data, cur_offset = Chunk.get_next_chunk_data(lol_file, cur_offset)
            reconstructed_data += chunk_data
            # If pointer is "null"
            if cur_offset == 0:
                break

        # Do verification step
        reconstructed_hash = hashlib.md5(reconstructed_data).digest()
        reconstructed_len = len(reconstructed_data)

        if reconstructed_hash == md5_bytes:
            logger.debug(
                f"Hash check ok (got {reconstructed_hash.hex()}, expected"
                f" {md5_bytes.hex()})"
            )
        else:
            raise RuntimeError(
                f"Hash check failed (got {reconstructed_hash.hex()}, but expected"
                f" {md5_bytes.hex()})"
            )

        # well, i'd be real impressed if it passes hash but not the length lol
        if reconstructed_len == len_bytes:
            logger.debug(
                f"Length check ok (got {reconstructed_len} bytes, exepcted"
                f" {len_bytes} bytes)"
            )
        else:
            raise RuntimeError(
                f"Length check failed (got {reconstructed_len} bytes, but exepcted"
                f" {len_bytes} bytes)"
            )

        return reconstructed_data


def get_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Creates and verifies .lol files.")
    parser.add_argument(
        "--input-file",
        default=DEFAULT_INPUT_FILE,
        type=Path,
        help="The file to convert into a .lol file.",
        required=True,
    )
    parser.add_argument(
        "--output-file",
        type=Path,
        default=None,
        help=(
            "The output path for the created .lol file. Defaults to the input file's"
            " stem with .lol instead of the original extension, if one existed."
        ),
    )
    parser.add_argument(
        "--seed",
        default=None,
        help=(
            "The random seed used to generate chunks. By default, a random seed"
            " is selected each time; you can pass a seed to make .lol generation"
            " deterministic."
        ),
    )
    parser.add_argument(
        "--min-chunk-size",
        "-l",
        type=int,
        default=DEFAULT_MIN_RANDOM_CHUNK_SIZE,
        help=(
            "The minimum length (in bytes) of the `data` field in an individual chunk,"
            " inclusive."
        ),
    )
    parser.add_argument(
        "--max-chunk-size",
        "-u",
        type=int,
        default=DEFAULT_MAX_RANDOM_CHUNK_SIZE,
        help=(
            "The maximum length (in bytes) of the `data` field in an individual chunk,"
            " exclusive."
        ),
    )

    return parser.parse_args()


def main(args: argparse.Namespace) -> None:
    if args.seed is not None:
        logger.debug(f"Using seed {args.seed}")
    else:
        logger.debug(f"No seed specified, using random seed")

    # Parse file into chunks, then pass them into LOLFile's method to convert it
    # to the "random" format
    raw_chunks = LOLFile.get_raw_chunks_from_file(
        args.input_file, args.min_chunk_size, args.max_chunk_size, args.seed
    )
    lol_file = LOLFile.from_chunks(raw_chunks, args.seed)

    # If no output file has been defined, just use the input filepath but with
    # .lol instead
    if not args.output_file:
        args.output_file = args.input_file.with_suffix(".lol")

    logger.info(f"Writing resulting file to {args.output_file}")

    # Write back out to file
    with open(args.output_file, "wb") as fp:
        fp.write(lol_file)

    # Reconstruct the file as a sanity check
    with open(args.output_file, "rb") as fp:
        try:
            reconstructed_data = LOLFile.undo_lol_file(fp.read())
        except:
            logger.exception("File-internal verify step failed")
            raise

    # Throw out the reconstructed file, just to verify. The default path is just
    # the input path with "-reconstructed" added onto its stem. Note that I use
    # with_name instead of with_stem for compatibility reasons.
    reconstructed_path = args.input_file.with_name(
        args.input_file.stem + "-reconstructed" + args.input_file.suffix
    )
    with open(reconstructed_path, "wb") as fp2:
        fp2.write(reconstructed_data)

    # Check if the file is still the same way going out as it was coming in
    with open(Path(args.input_file), "rb") as fp:
        old_hash = hashlib.md5(fp.read()).digest()

    new_hash = hashlib.md5(reconstructed_data).digest()
    if new_hash == old_hash:
        logger.info("OK")
    else:
        logger.error("Hash verify step failed")
        exit(1)


if __name__ == "__main__":
    # Parse arguments
    args = get_args()

    main(args)

File: re/lol/lol.c
/**
 * @brief The source of the binary to be reversed.
 * @author Lloyd Gonzales (lgonzalesna@gmail.com)
 * 
 * This is the "reverse" of lol.py, which reconstructs the file in-memory and then
 * verifies that it matches the md5 contained in the file.
 * 
 * Naturally, because the file gets fully reconstructed, there's a lot of options
 * to recover the file. The original intent was that you're supposed to rebuild the
 * parser yourself by looking at the code in a decompiler and figuring out what the
 * file format is. But you could almost certainly do something like patching out the
 * binary to loop right after the file is reconstructed (but before the program exits),
 * dumping its memory, then looking for the file in its heap.
 * 
 * References:
 * - https://stackoverflow.com/questions/10324611/how-to-calculate-the-md5-hash-of-a-large-file-in-c
 */

#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#if defined(__APPLE__)
#  define COMMON_DIGEST_FOR_OPENSSL
#  include <CommonCrypto/CommonDigest.h>
#  define SHA1 CC_SHA1
#else
#  include <openssl/md5.h>
#endif

bool g_is_little_endian = false;

typedef struct {
    uint64_t file_length;
    unsigned char file_md5[MD5_DIGEST_LENGTH];
    char* data;
} lol_file_t;


/**
 * @brief Calculate the md5 sum of an array of bytes.
 * 
 * \param [in, out] digest The calculated md5 digest.
 * \param [in] data The data to calculate the md5 sum of.
 * \param [in] data_len The length of the data to calculate the md5 sum over.
 */
void calculate_md5(unsigned char digest[MD5_DIGEST_LENGTH], unsigned char data[], uint64_t data_len){
    MD5_CTX mdContext;

    MD5_Init(&mdContext);
    MD5_Update (&mdContext, data, data_len);
    MD5_Final (digest, &mdContext);
}

/**
 * \brief Print out an md5 digest to hex.
 * 
 * \param digest The md5 digest to print out. Should be exactly 16 bytes.
 */
void print_md5(unsigned char digest[MD5_DIGEST_LENGTH]){
    char out[33] = {0};
    for (int n = 0; n < 16; ++n) {
        snprintf(&(out[n*2]), 16*2, "%02x", (unsigned int)digest[n]);
    }
    printf("%s\n", out);
}

/**
 * @brief Check that two MD5 digests are equal.
 * 
 * this really only exists because these aren't null-terminated so strcmp()
 * doesn't work, but i'd also like to keep the clarity of MD5_DIGEST_LENGTH around
 * 
 * \param [in] digest_1 The first digest to compare.
 * \param [in] digest_2 The second digest to compare.
 * \retval true if the digests are equal
 * \retval false if the digest are not equal
 */
bool compare_md5(unsigned char digest_1[MD5_DIGEST_LENGTH], unsigned char digest_2[MD5_DIGEST_LENGTH]){
    for(int i = 0; i < MD5_DIGEST_LENGTH; i++){
        if(digest_1[i] != digest_2[i]){
            return false;
        }
    }

    return true;
}

/**
 * \brief If necessary, convert a BE uint32 to the machine's endianness.
 * 
 * Ref: https://stackoverflow.com/questions/2182002/how-to-convert-big-endian-to-little-endian-in-c-without-using-library-functions
 * 
 * \param [in] val A big-endian uint32.
 * \return uint32_t The correctly-interpreted value.
 */
uint32_t interpret_uint32(uint32_t val){
    if (g_is_little_endian){
        val = ((val << 8) & 0xFF00FF00 ) | ((val >> 8) & 0xFF00FF ); 
        return (val << 16) | (val >> 16);
    }

    return val;
}

/**
 * \brief If necessary, convert a BE uint64 to the machine's endianness.
 * 
 * Ref: https://stackoverflow.com/questions/2182002/how-to-convert-big-endian-to-little-endian-in-c-without-using-library-functions
 * 
 * \param [in] val A big-endian uint64.
 * \return uint64_t The correctly-interpreted value.
 */
uint64_t interpret_uint64(uint64_t val){
    if (g_is_little_endian){
        val = ((val << 8) & 0xFF00FF00FF00FF00ULL ) | ((val >> 8) & 0x00FF00FF00FF00FFULL );
        val = ((val << 16) & 0xFFFF0000FFFF0000ULL ) | ((val >> 16) & 0x0000FFFF0000FFFFULL );
        return (val << 32) | (val >> 32);
    }

    return val;
}

int main(int argc, char* argv[]){
    // Original file
    char* filename;
    FILE *fp;

    // Lolfile
    lol_file_t lolfile = {0};

    // Reconstructed info
    unsigned char* reconstructed_data;
    uint64_t cur_pos = 0;
    unsigned char reconstructed_md5[MD5_DIGEST_LENGTH];

    if (argc == 2){
        filename = argv[1];
    } else {
        printf("Usage: %s file.lol\n", argv[0]);
        exit(1);
    }

    // Check for endianness.
    int n = 1;
    // little endian if true
    if(*(char *)&n == 1){
        g_is_little_endian = true;
    }

    fp = fopen(filename, "rb");

    // Get file length.
    fread(&lolfile.file_length, sizeof(lolfile.file_length), 1, fp);
    lolfile.file_length = interpret_uint64(lolfile.file_length);
    
    // Get final MD5.
    fread(&lolfile.file_md5, sizeof(lolfile.file_md5), 1, fp);

    // Create the in-memory file.
    reconstructed_data = malloc(sizeof(unsigned char) * lolfile.file_length);

    assert(reconstructed_data != NULL);

    // At this point, fp is at the start of the first chunk. All we really
    // need to do now is start reading chunks off and seek back to the absolute
    // position whenever necessary.
    while(true){
        uint32_t chunk_length;
        uint64_t next_chunk;

        // Get length of chunk data.
        fread(&chunk_length, sizeof(chunk_length), 1, fp);
        chunk_length = interpret_uint32(chunk_length);

        // Check that the ending position of reading in this chunk won't overrun
        // the internal file's boundaries.
        if (cur_pos + chunk_length > lolfile.file_length){
            // Maybe this'll be a nice hint?
            printf("File is longer than expected, final pointer may be missing.\n");
            exit(1);
        }

        // Read that many into reconstructed_data, plus whatever the
        // current offset is.
        fread(reconstructed_data + cur_pos, chunk_length, 1, fp);
        cur_pos += chunk_length;

        // Figure out where the next chunk is.
        fread(&next_chunk, sizeof(next_chunk), 1, fp);
        next_chunk = interpret_uint64(next_chunk);

        // If the next chunk is "null", then break. Else, update fp to
        // that position.
        if (next_chunk == 0){
            // File has been reconstructed.
            break;
        } else {
            fseek(fp, next_chunk, SEEK_SET);
        }
    }

    // Check that the md5 sum is identical
    calculate_md5(reconstructed_md5, reconstructed_data, lolfile.file_length);
    if(compare_md5(reconstructed_md5, lolfile.file_md5)){
        printf("OK\n");
        exit(0);
    }else{
        printf("Check failed\n");
        exit(1);

        // Could leave these in to make it more obvious what's happening
        // print_md5(reconstructed_md5);
        // print_md5(lolfile.file_md5);
    }
}